tests/scripts/all.sh

```diff
@@ -1120,6 +1120,70 @@ component_test_cmake_out_of_source () {
+    #scripts/config.pl set MBEDTLS_NO_PLATFORM_ENTROPY Need some entropy
+    scripts/config.pl unset MBEDTLS_CIPHER_MODE_CFB
+    scripts/config.pl unset MBEDTLS_CIPHER_MODE_OFB
+    scripts/config.pl unset MBEDTLS_CIPHER_MODE_CTR
+    scripts/config.pl unset MBEDTLS_CIPHER_PADDING_ONE_AND_ZEROS
+    scripts/config.pl unset MBEDTLS_CIPHER_PADDING_ZEROS_AND_LEN
+    scripts/config.pl unset MBEDTLS_CIPHER_PADDING_ZEROS
+    scripts/config.pl unset MBEDTLS_CIPHER_MODE_XTS
+    scripts/config.pl unset MBEDTLS_ECP_DP_SECP192R1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_SECP224R1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_SECP521R1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_SECP192K1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_SECP224K1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_SECP256K1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_BP256R1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_BP384R1_ENABLED
+    scripts/config.pl unset MBEDTLS_ECP_DP_BP512R1_ENABLED
+    scripts/config.pl unset MBEDTLS_PK_PARSE_EC_EXTENDED
+    scripts/config.pl unset MBEDTLS_AESNI_C
+    scripts/config.pl unset MBEDTLS_ARC4_C
+    scripts/config.pl unset MBEDTLS_BLOWFISH_C
+    scripts/config.pl unset MBEDTLS_CAMELLIA_C
+    scripts/config.pl unset MBEDTLS_DES_C
+    scripts/config.pl unset MBEDTLS_DHM_C
+    scripts/config.pl unset MBEDTLS_GENPRIME
+    scripts/config.pl unset MBEDTLS_MD5_C
+    scripts/config.pl unset MBEDTLS_PADLOCK_C
+    scripts/config.pl unset MBEDTLS_PEM_WRITE_C
+    scripts/config.pl unset MBEDTLS_PKCS5_C
+    scripts/config.pl unset MBEDTLS_PKCS12_C
+    scripts/config.pl unset MBEDTLS_RIPEMD160_C
+    scripts/config.pl unset MBEDTLS_SHA1_C
+    scripts/config.pl unset MBEDTLS_XTEA_C
+    scripts/config.pl set MBEDTLS_CMAC_C
+    scripts/config.pl set MBEDTLS_AES_ROM_TABLES
+    scripts/config.pl unset MBEDTLS_X509_RSASSA_PSS_SUPPORT
+    scripts/config.pl unset MBEDTLS_X509_CSR_PARSE_C
+    scripts/config.pl unset MBEDTLS_X509_CREATE_C
+    scripts/config.pl unset MBEDTLS_X509_CRT_WRITE_C
+    scripts/config.pl unset MBEDTLS_X509_CSR_WRITE_C
+    scripts/config.pl unset MBEDTLS_KEY_EXCHANGE_DHE_PSK_ENABLED
+    scripts/config.pl unset MBEDTLS_KEY_EXCHANGE_RSA_PSK_ENABLED
+    scripts/config.pl unset MBEDTLS_KEY_EXCHANGE_RSA_ENABLED
+    scripts/config.pl unset MBEDTLS_KEY_EXCHANGE_DHE_RSA_ENABLED
+    scripts/config.pl unset MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED
+    scripts/config.pl unset MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED
+    scripts/config.pl unset MBEDTLS_SSL_FALLBACK_SCSV
+    scripts/config.pl unset MBEDTLS_SSL_CBC_RECORD_SPLITTING
+    scripts/config.pl unset MBEDTLS_SSL_PROTO_TLS1
+    scripts/config.pl unset MBEDTLS_SSL_PROTO_TLS1_1
+    scripts/config.pl unset MBEDTLS_SSL_TRUNCATED_HMAC
+    scripts/config.pl unset MBEDTLS_PLATFORM_TIME_TYPE_MACRO
+    scripts/config.pl set MBEDTLS_MPI_MAX_SIZE     512
+    make CC=gcc CFLAGS='-Werror -O1' all test
```

Comment:
Should we also add `-Wstack-usage=2048 -Werror` to ensure we don't overflow the Mbed OS default main thread stack?

We should also use `-Os`, if that's what Mbed OS will default to using.

Reply:
Gah, seems we can't enforce that stack limit until we can fit within that stack limit.

```
$ make CFLAGS:="-Os -Wstack-usage=2048 -Werror" check
  CC    bignum.c
bignum.c: In function ‘mbedtls_mpi_exp_mod’:
bignum.c:1926:5: error: stack usage is 3312 bytes [-Werror=stack-usage=]
 int mbedtls_mpi_exp_mod( mbedtls_mpi *X, const mbedtls_mpi *A,
     ^~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
Makefile:150: recipe for target 'bignum.o' failed
make[2]: *** [bignum.o] Error 1
Makefile:158: recipe for target 'libmbedcrypto.a' failed
make[1]: *** [libmbedcrypto.a] Error 2
Makefile:19: recipe for target 'lib' failed
make: *** [lib] Error 2
```
What stack size should we test that we don't exceed and recommend Mbed OS users configure if they want to use Mbed TLS? We appear to fit within 6000 bytes.

```
  Gen   test_suite_version.c
  CC    test_suite_version.c
suites/host_test.function: In function ‘execute_tests’:
suites/host_test.function:494:5: error: stack usage might be 5968 bytes [-Werror=stack-usage=]
 int execute_tests( int argc , const char ** argv )
     ^~~~~~~~~~~~~
```

Reply:
Stack consumption in `mbedtls_mpi_exp_mod` increases exponentially with `MBEDTLS_MPI_WINDOW_SIZE`. It's a space/time compromise, with ~1.5kB difference in stack usage between the min and max value on a 32-bit machine. Maybe we could reduce this value in Mbed OS: currently it uses the best-performance value except on `TARGET_STM32F439xI`.

The library and the test suite are written in a very different style when it comes to stack consumption. `execute_tests` uses a `char buf[5000]` to store a line of the `.data` file. I think we should `-Wstack-usage` only for the library, not for the test suite.

Reply:
I agree we should run with `-Wstack-usage` only for the library for this PR.

In the future, we should build the tests as they'd be used on target (so, likely excluding host_test.function): this enables us to catch when on-target testing would run into stack issues. Some test suites allocate unnecessarily large buffers on the stack. We should have a regression test in place to ensure issues like what https://github.com/ARMmbed/mbed-crypto/pull/278 fixes don't rear their heads again, but adding such tests in this PR is out of scope.